### PR TITLE
Enable token flag for rpc auth

### LINKED ISSUE #2009: Improve ergonomics of token handling in the forest client
**Issue summary**

Forest uses a [JSON Web Token](https://jwt.io/) to manage access to the API endpoints. For example, querying the status of the forest node only requires `read` permissions but exporting the private keys of a wallet requires `admin` rights.

The tokens are generated by the daemon and printed to stdout. Using the tokens in the client is not easy, though. Right now it can only be done by setting the `FULLNODE_API_INFO` environment variable. We also have a `--token` flag but it is completely ignored(!).

To be done:
- Either remove the `--token` flag or use it to pass the token to the client.
- Consider reading the token from the forest keystore. The keystore may or may not be password protected.

**Other information and links**

https://github.com/ChainSafe/forest/blob/5e7150b44f64b064a3d38e51ff02e45719375b91/forest/src/cli/client.rs#L20
https://github.com/ChainSafe/forest/blob/5e7150b44f64b064a3d38e51ff02e45719375b91/node/rpc-api/src/lib.rs#L140
https://github.com/ChainSafe/forest/blob/5e7150b44f64b064a3d38e51ff02e45719375b91/node/rpc-client/src/lib.rs#L44-L59

### PR DESCRIPTION
**Pull request by sudo-shashank**
**Summary of changes**
Changes introduced in this pull request:
- Enable use of `--token` flag to pass jwt auth token to rpc method `call` 

**Reference issue to close (if applicable)**

Closes #2009 

**Other information and links**

### CODE REVIEW

**Comment by lemmih** [comment]
Once it's ready for review, it should no longer be a draft PR.

File: blockchain/consensus/deleg_cns/README.md [hunk 1]
```diff
@@ -360,20 +360,17 @@ Let's see what happens if we create a new wallet. We don't need to do this, becu
 
 ```console
 $ forest --token $JWT_TOKEN wallet new bls
```
Comment (hanabi1224) [inline blockchain/consensus/deleg_cns/README.md:362]: It should now be `forest-cli`

File: blockchain/consensus/deleg_cns/README.md [hunk 2]
```diff
@@ -360,20 +360,17 @@ Let's see what happens if we create a new wallet. We don't need to do this, becu
...
 using the `FULLNODE_API_INFO` env var, which is expected to be a concatenation of the JWT token and the API multiaddress.
 
 ```bash
 export FULLNODE_API_INFO=$JWT_TOKEN:/ip4/127.0.0.1/tcp/1234/http
 ```
 
-Check again:
-
 ```console
 $ forest wallet new bls
```
Comment (hanabi1224) [inline blockchain/consensus/deleg_cns/README.md:375]: ditto

Comment (sudo-shashank) [reply blockchain/consensus/deleg_cns/README.md]: corrected

File: blockchain/consensus/deleg_cns/README.md (hunk 1 above)
Comment (sudo-shashank) [reply blockchain/consensus/deleg_cns/README.md]: corrected

**Comment by elmattic** [comment]
I'm not able to use `--token` flag with `forest-cli`, daemon answered with:
`ERROR forest_rpc::rpc_util                > RPC call error: Unknown error after making RPC call. Code: 0. Error: "Missing required claim: exp"`

Is this normal? Just wanted to test the various commands, like creating a wallet.

**Comment by LesnyRumcajs** [comment]
@elmattic  Ya, I also wasn't able to figure out quickly what's the workflow here. That's why I asked for some piece of documentation for dummies.

**Comment by lemmih** [comment]
> I'm not able to use `--token` flag with `forest-cli`, daemon answered with: `ERROR forest_rpc::rpc_util > RPC call error: Unknown error after making RPC call. Code: 0. Error: "Missing required claim: exp"`
> 
> Is this normal? Just wanted to test the various commands, like creating a wallet.

I might have fixed something related to that recently. I'll update the branch and test.

**Comment by elmattic** [comment]
> > I'm not able to use `--token` flag with `forest-cli`, daemon answered with: `ERROR forest_rpc::rpc_util > RPC call error: Unknown error after making RPC call. Code: 0. Error: "Missing required claim: exp"`
> > Is this normal? Just wanted to test the various commands, like creating a wallet.
> 
> I might have fixed something related to that recently. I'll update the branch and test.

Hmm thanks but didn't help. Will try to see test generating a new Token.

**Comment by elmattic** [comment]
Hey PR looks good but what about the documentation part, as hinted by Hubert?

**Comment by sudo-shashank** [comment]
The tokens are generated by the daemon and printed to stdout and usage of `--token` flag has been updated in Readme 
https://github.com/ChainSafe/forest/pull/2068/files#diff-a486afe20d3b13540b48fda776d2ef888f246e5496fd4b3cd992b372bd92badb

**Comment by LesnyRumcajs** [comment]
@sudo-shashank It'd be great to update the high-level README. The one in `deleg_cns` is a bit hidden and niche.
https://github.com/ChainSafe/forest/blob/main/README.md?plain=1#L199-L214

File: README.md [hunk 3]
```diff
@@ -212,6 +212,11 @@ Forest developers will prepend this variable to CLI commands over using `export`
 FULLNODE_API_INFO="..." forest auth api-info -p admin
```
Comment (LesnyRumcajs) [inline README.md:214]: It'd be great to update it to the new CLI binary, i.e. `FULLNODE_API_INFO="..." forest-cli auth api-info -p admin`

File: utils/auth/src/lib.rs [hunk 4]
```diff
@@ -64,6 +64,7 @@ pub fn create_token(perms: Vec<String>, key: &[u8]) -> JWTResult<String> {
 pub fn verify_token(token: &str, key: &[u8]) -> JWTResult<Vec<String>> {
     let mut validation = jsonwebtoken::Validation::new(jsonwebtoken::Algorithm::default());
     validation.validate_exp = false;
+    validation.required_spec_claims.remove("exp");
```
Comment (LesnyRumcajs) [inline utils/auth/src/lib.rs:67]: I'm not sure it shouldn't be the other way around, i.e. adding this expiration to the token creation and validating it. Thoughts?

File: README.md (hunk 3 above)
Comment (sudo-shashank) [reply README.md]: corrected

File: utils/auth/src/lib.rs (hunk 4 above)
Comment (sudo-shashank) [reply utils/auth/src/lib.rs:67]: I think adding `exp` to token payload is a better way to handle this
but right now admin token is printed out to stdout when daemon starts.
I am not sure if we have an option to request for a new admin token
another way would be to print out a new admin token once the previous one expires 

also what should be the expiration time for this token?

Comment (LesnyRumcajs) [reply utils/auth/src/lib.rs:67]: We can generate a new admin token every time we start `forest`. To my understanding, it should not invalidate the previous ones. Only the signature and the expiration time would be different, but they would all be signed by the same private key.

Expiration time could be, e.g. 60 days, optimally configurable via a config flag. We can do it all in a separate task. If you want to proceed this way, could you please create an issue to implement proper expiration handling (and creation)?

Comment (sudo-shashank) [reply utils/auth/src/lib.rs:67]: yeah sounds good
I'll create a new issue for this and work on it

Comment (LesnyRumcajs) [reply utils/auth/src/lib.rs:67]: Alright, please link it here and then we can merge this one.

**Review by lemmih — APPROVED** [review]

**Pull request merged**